File: 2020/11.rb
DXY = [-1, 0, 1].freeze

def run(lines, p2)
	temp = ""
	(0..lines.size-1).each do |row|
		(0..lines[row].size-1).each do |col|
			temp << nearby(row, col, lines, p2)
		end
		yield temp
		temp = ""
	end
	return temp
end

def nearby(row, col, lines, p2)
	seat = lines[row][col]
	return seat if seat == "."
	count = 0
	
	9.times do |i|
		diff = [DXY[i / 3], DXY[i % 3]]
		r = row + diff[0]
		c = col + diff[1]
		inRange = lambda { |r, c| r.between?(0,92) && c.between?(0,94)} #saving space
		if !p2
			if inRange.call(r, c)
				count += 1 if lines[r][c] == "#"
			end
		else
			while inRange.call(r, c)
				count += 1 if lines[r][c] == "#"
				break if lines[r][c] == "L" || lines[r][c] == "#"
				r += diff[0]
				c += diff[1]
			end
		end
	end

	max = p2 ? 6: 5 #+1 because I counted the current seat
	case seat
	when "L"
		return count == 0 ? "#": "L"
	else
		return count < max ? "#": "L" 
	end
end

lines = File.open("11.txt").read.split("\n")
p2 = false
2.times do |i|
	nSeats, temp = lines[0..-1], []
	while nSeats != temp
		temp = nSeats[0..-1]
		nSeats = []
		run(temp, p2) {|row| nSeats << row}
	end
	puts nSeats.map {|row| row.count "#"}.sum
	p2 = !p2
end

File: 2020/11.py
lines = open("11.txt").read().split("\n")
grid = [[*line] for line in lines]

# Count the number of adjacent seats that are full
def count_adj(grid, r, c):
    count = 0
    for i in range(-1, 2):
        for j in range(-1, 2):
            if i != 0 or j != 0:
                if r + i >= 0 and r + i < len(grid) and c + j >= 0 and c + j < len(grid[0]):
                    if grid[r + i][c + j] == "#": count += 1
    return count

# Count the number of visible seats that are full. A seat is visible from a given seat if there is
# only floor between them
def count_vis(grid, r, c):
    count = 0
    for i in range(-1, 2):
        for j in range(-1, 2):
            if i != 0 or j != 0:
                k, l = 1, 1
                while r + i * k >= 0 and r + i * k < len(grid) and c + j * l >= 0 and c + j * l < len(grid[0]):
                    if grid[r + i * k][c + j * l] == "#": count += 1; break
                    elif grid[r + i * k][c + j * l] == "L": break
                    k += 1; l += 1
    return count

def run(grid, count_func, adjacent_full_limit):
    while True:
        temp_grid = [[c for c in row] for row in grid]
        for r in range(len(grid)):
            for c in range(len(grid[r])):
                if grid[r][c] == "L":
                    adj_count = count_func(grid, r, c)
                    temp_grid[r][c] = "#" if adj_count == 0 else "L"
                elif grid[r][c] == "#":
                    adj_count = count_func(grid, r, c)
                    temp_grid[r][c] = "#" if adj_count < adjacent_full_limit else "L"
                else: temp_grid[r][c] = "."
        if temp_grid == grid: break
        grid = [[c for c in row] for row in temp_grid]
    print (sum(row.count("#") for row in grid))

# Part 1: ran in to issues with array copying as I always do. Accidentally reversed > with < as
run(grid, count_adj, 4)
# Part 2: took a couple of tries to get right, but it's actually pretty easy to just add count_vis
run(grid, count_vis, 5)
